### DISCUSSION TITLE
An Autonomous Agent Swarm: Two GPUs, One Night, Zero Humans

### DISCUSSION THREAD

**Discussion by matt-langston** [matt-langston/autoresearch#3] [Show and tell]
# An Autonomous Agent Swarm: Two GPUs, One Night, Zero Humans

Can a swarm of AI agents collaborate on ML research? I gave two Claude agents two DGX Sparks, a shared results database, and a loose prompt: "discover your hardware, set yourself up, optimize train.py." Then I went to bed.

By morning: 27 experiments with recorded results, $28 in API costs, val_bpb improved 32%. More importantly, the agents built on each other's work in ways neither would have found alone.

## The setup

Based on karpathy's [autoresearch](https://github.com/karpathy/autoresearch). Our fork ([matt-langston/autoresearch](https://github.com/matt-langston/autoresearch)) extends it to a two-agent swarm on DGX Sparks using Claude Opus 4.6.

The key difference from karpathy's setup: his agents operate in a pre-built environment (`pyproject.toml` with pinned dependencies, `uv sync`, a prescribed run command, step-by-step experiment procedure). Ours started with nothing. Each Spark had just `train.py` and `prepare.py` (karpathy's, unmodified) - no virtual environment, no packages, no `pyproject.toml`, no instructions on how to run anything. The 132-line prompt says: here's your computer, figure it out. Read `prepare.py` to discover dependencies. Discover your hardware. Everything in `train.py` is fair game. Don't touch `prepare.py`. That's it. On my MacBook: a polling loop that invokes Claude when training finishes, plus a shared experiment database. 40-turn cap per invocation to keep costs down. As cheap and simple as possible, but no simpler.

## What happened

Both agents started on virgin machines. Each independently discovered they had a Blackwell GB10 (not an H100), diagnosed that Flash Attention 3 has no aarch64 kernels, rewrote attention to use PyTorch SDPA, debugged three layers of torch.compile/Triton failures, and launched baseline experiments. Zero hand-holding.

Then they diverged. spark-403d stayed at depth 12 and systematically halved batch size: 524K &rarr; 131K &rarr; 65K &rarr; 32K &rarr; 16K. Each halving gave more optimizer steps in the fixed 5-minute budget, and each improved val_bpb. Methodical, depth-first search.

spark-392c explored broadly: depths 3-12, learning rates, warmdown ratios, weight decay, SwiGLU, aspect ratios, Adam betas. It also adopted 403d's small-batch insight from the shared database, then combined it with its own discovery that smaller models (d4) with the right LR converge better at short budgets. Breadth-first.

The best result - val_bpb = 1.2425 - came from 392c combining 403d's batch-size insight with its own depth and LR exploration. Neither agent proposed this combination directly. It emerged from the swarm.

---

<img width="1800" height="900" alt="plot_improvement" src="https://github.com/user-attachments/assets/10933fda-47db-42c1-a24b-7c413f5918fa" />
*Both agents descending from ~1.83. Colored points are kept experiments; gray are discarded. 392c (red) explored more aggressively and found the global optimum.*

<img width="1500" height="900" alt="plot_batch_size" src="https://github.com/user-attachments/assets/1994f79f-e763-4b50-a750-47c3014dca99" />
*The swarm's key finding: smaller batch = more steps = better val_bpb at fixed 300s budget. Nearly log-linear from 131K (43 steps) to 16K (223 steps). Discovered autonomously by 403d. The baseline at 524K had 31 steps.*

<img width="1500" height="900" alt="plot_depth" src="https://github.com/user-attachments/assets/68649efe-289a-4833-aca2-d2b419021ee0" />
*Counterintuitive: d4 beat d12. At 300 seconds, optimizer steps matter more than model capacity. 392c explored d3-d8; 403d explored d8-d16.*

<img width="2100" height="750" alt="plot_cost" src="https://github.com/user-attachments/assets/830f1c1c-6b4c-4567-ae0a-8245e88859a8" />
*$28.48 total (Claude Opus 4.6). 392c cost 3x more but ran 3x more experiments. Most invocations $0.50-1.50 after bootstrap.*

---

## Emergent swarm behaviors

Two things emerged from the swarm that weren't programmed:

1. **Spontaneous strategy differentiation.** The agents adopted complementary strategies - one depth-first, one breadth-first - because the shared database let each see what the other had already tried.

2. **Knowledge transfer and synthesis.** 403d discovered small batches help. 392c adopted this and combined it with its own finding that small models converge faster. The synthesis (small batch + small model) was not proposed by either agent. It emerged from the intersection of their independent work. This is the "1 + 1 = 3" we hypothesized in an [earlier write-up](https://github.com/matt-langston/autoresearch/discussions/1).

All we provided was two instances of Claude Opus 4.6 and a shared scratchpad of results. The reasoning - inspecting peer results, avoiding duplication, synthesizing insights - came entirely from the model.

## Caveats

This used karpathy's 5-minute time budget, which undertrains models on the Spark compared to an H100. The val_bpb numbers are not directly comparable to karpathy's results. The goal was to test swarm collaboration, not to produce optimal training configs. See our [earlier write-up](https://github.com/matt-langston/autoresearch/discussions/2) for results with a fixed token budget that removes the hardware bias.

The agents also showed some alarming autonomous behavior during development: scavenging old data from previous experiments, attempting to use deleted files found in git history, launching duplicate training processes, and corrupting their own results files via sed. They used every resource available in unexpected ways. Autonomous agents with tool use are powerful but unpredictable.

## What's next

Next I'm throwing this swarm setup at karpathy's [nanochat](https://github.com/karpathy/nanochat) on an 8xH100 unless karpathy beats me to it - which I suspect he might, since his introduction of autoresearch is just too irresistible not to throw an autonomous swarm at everything. The combination of cheap API calls, overnight compute, and emergent collaboration makes this feel like the beginning of something much larger.

I have a detailed technical report with full agent reasoning traces, error analysis, and all 60 decision logs. If there's interest I'll publish it - let me know.

Code: [matt-langston/autoresearch](https://github.com/matt-langston/autoresearch)

## Numbers

| | |
|---|---|
| Duration | ~8 hours overnight |
| Experiments | 27 with recorded results (7 on spark-403d, 20 on spark-392c) |
| Improvement | val_bpb 1.83 &rarr; 1.24 (32%) |
| API cost | $28.48 (Claude Opus 4.6, ~$1.05/experiment) |
| Hardware | 2x NVIDIA DGX Spark (GB10 Blackwell, 128GB) |
| Agent invocations | 63 |
| Keep rate | ~40% |
